### PR TITLE
bug: ptreeToRapidJSON does not handle base cases

### PR DESCRIPTION
**Pull request by muffins**
I think this'll work >.>

### CODE REVIEW

File: osquery/database/tests/database_tests.cpp [hunk 1]
```diff
@@ -130,7 +130,15 @@ static void ptree_to_json(const pt::ptree& pt_in, const JSON& rj_in) {
...
-  EXPECT_EQ(rj_in.doc(), rj_out.doc());
+  std::string ptree_converted{""};
+  std::string rjin_expected{""};
+
+  auto s = rj_in.toString(rjin_expected);
+  EXPECT_TRUE(s.ok());
+  s = rj_out.toString(ptree_converted);
+  EXPECT_TRUE(s.ok());
+
+  EXPECT_EQ(ptree_converted, rjin_expected);
```
Comment (fmanco) [inline osquery/database/tests/database_tests.cpp:141]: Why compare the strings instead of the objects?

File: osquery/database/tests/database_tests.cpp [hunk 2]
```diff
@@ -149,18 +157,23 @@ TEST_F(DatabaseTests, test_ptree_to_rj_object) {
   pt::ptree pt_in;
   pt_in.put("k", "v");
 
-  JSON rj_in = JSON::newObject();
-  rj_in.add("k", "v");
+  JSON rj_in = JSON::newArray();
+  JSON rj_obj = rj_in.newObject();
+  rj_obj.add("k", "v");
+  rj_in.push(rj_obj.doc(), rj_in.doc());
```
Comment (fmanco) [inline osquery/database/tests/database_tests.cpp:163]: Isn't this going to create `[ { "k" : "v" } ]`?

File: osquery/database/tests/database_tests.cpp [hunk 3]
```diff
@@ -149,18 +157,23 @@ TEST_F(DatabaseTests, test_ptree_to_rj_object) {
...
 
-  pt::ptree pt_in;
-  pt_in.put("", "a");
-  pt_in.put("", "b");
+  pt::ptree pt_in, child1, child2, children;
+  child1.put("", "a");
+  child2.put("", "b");
+
+  pt_in.add_child("", child1);
+  pt_in.add_child("", child2);
```
Comment (fmanco) [inline osquery/database/tests/database_tests.cpp:176]: Isn't this going to create a `{ "" : { "" : "a" } , "" : { "" : "b" } }` which should be converted to `[ ["a"] , ["b"] ]`?

File: osquery/database/database.cpp [hunk 4]
```diff
@@ -366,16 +366,31 @@ Status ptreeToRapidJSON(const std::string& in, std::string& out) {
     return Status(1, "Failed to parse JSON");
   }
 
+  if (tree.empty()) {
+    JSON().toString(out);
+    return Status();
+  }
+
   auto json = JSON::newArray();
```
Comment (fmanco) [inline osquery/database/database.cpp:374]: I don't think this works for instance for `{ "v1" : "k1" , "v2" : "k2" }`. This shouldn't be an array but rather an object.

File: osquery/database/database.cpp [hunk 5]
```diff
@@ -366,16 +366,31 @@ Status ptreeToRapidJSON(const std::string& in, std::string& out) {
...
-    rj::Document row;
-    if (row.Parse(ss.str()).HasParseError()) {
-      return Status(1, "Failed to serialize JSON");
+    JSON obj = json.newObject();
+
+    if (!t.second.empty()) {
+      // if the node is a subtree, we parse this directly with RJ
+      std::stringstream ss;
+      pt::write_json(ss, t.second);
+      if (obj.doc().Parse(ss.str()).HasParseError()) {
```
Comment (fmanco) [inline osquery/database/database.cpp:382]: Isn't parsing ptree generated json with rapidjson dangerous? I think we need to parse everything manually, maybe recursively.

File: osquery/database/database.cpp (hunk 4 above)
Comment (muffins) [reply osquery/database/database.cpp:374]: Sure, I think the reason we had done this was largely because this upgrade function wasn't intended to perfectly convert a ptree to an RJ, it was just meant to handle the edge cases we were seeing in the upgrade path where lists in ptrees were not correct list json types.

File: osquery/database/database.cpp (hunk 5 above)
Comment (muffins) [reply osquery/database/database.cpp:382]: Yeah see the above comment. This function wasn't intended to be a perfect conversion, it was just meant to handle the upgrade cases to get us past that error we were seeing with the object not being a list. I think we might be able to get away with more limited functionality on this function, as it was mostly designed to convert the query data contents, which ended up being something that RJ could read in from ptree written out as a json string.

File: osquery/database/tests/database_tests.cpp (hunk 1 above)
Comment (muffins) [reply osquery/database/tests/database_tests.cpp:141]: It's my first impression that RJ doesn't have an `operator=` defined. We can add one if we like, but comparing these two objects directly wont give us the comparison we'd like, as it's attempting to see if the class object memory locations are the same, which they wont be. I think what we're wanting here is to see if the json data itself is the same, for which we'd need to access the string values. Doing the correct comparison as we had it will never pass as the class objects will have different memory locations, thus failing the test.

File: osquery/database/tests/database_tests.cpp (hunk 2 above)
Comment (muffins) [reply osquery/database/tests/database_tests.cpp:163]: Yeah see the notes above. The change in the json structure for the unit tests was meeting the updated "upgrade" function logic halfway. Again, we can change all of these things, but building a full ptree -> RJ function might take quite a bit of time and be tricky, and I was only intending to unblock us. Further I'm not 100% convinced that we need a full conversion function? But I definitely could be wrong on that.

File: osquery/database/tests/database_tests.cpp (hunk 3 above)
Comment (muffins) [reply osquery/database/tests/database_tests.cpp:176]: Same as above.

**Comment by muffins** [comment]
I'll take a look at this again today, and see if I can whip up the fully functionalty ptree -> RJ code, however in my time messing with this yesterday it looked messy, and I think there's quite a few edge cases here that'll screw with us. While I agree that's the safer way forward, I think having ptree write out a json string and having RJ serialize that in was working well for us, we simply had a few edge cases that were breaking things, such as empty json objects like you have in your first unit tests. Lemme know what you think, as I honestly just want good path forward for this as we're blocked on the 3.2.3 release.

**Comment by fmanco** [comment]
@muffins Let's kill this one too.
